File: Python/283.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        # Two pointer method
        if len(nums) < 2: return

        slow = 0

        for fast in range(len(nums)):
            if nums[slow] == 0 and nums[fast] != 0:
                nums[slow], nums[fast] = nums[fast], nums[slow]
                slow += 1

            # If nums[slow] is not zero -> no swaps happening
            if nums[slow] != 0:
                slow += 1
                

File: CPP/283.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {

        if (nums.size() < 2) return;

        size_t slow = 0;

        for (size_t index = 0; index < nums.size(); ++index){
            if (nums[slow] == 0 && nums[index] != 0){
                // Swap happening
                swap(nums[slow], nums[index]);
            }

            if (nums[slow] != 0) slow++;
        }
    }
};

class Solution {
public:
    void moveZeroes(vector<int>& nums) {

        int vecSize = nums.size();
        // two pointer - find non-zero and insert it in slow pos.
        size_t slow = 0, fast = 0;
        for (; fast < vecSize; ++fast){
            if (nums[fast] != 0){
                nums[slow] = nums[fast];
                ++slow;
            }
        }
        // slow -> starts of zeros
        for (; slow < vecSize; ++slow){
            nums[slow] = 0;
        }


    }
};
